torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -730,6 +750,75 @@ void Reducer::finalize_backward() {
     local_used_work_->wait();
   }
   local_used_maps_reduced_ = false;
+
+  // Rebuilt bucket if this is the first time to rebuilt
+  if (!rebuilt_tensors_.empty()) {
+    rebuildBuckets();
+  }
+}
+
+void Reducer::sync_bucket_indices(
+    std::vector<std::vector<size_t>>& bucket_indices) {
+  int64_t broadcast_bucket_size = DEFAULT_BROADCAST_BUCKET_BYTES;
+  auto num_buckets = bucket_indices.size();
+  int64_t total_size = 0;
+  for (size_t i = 0; i < num_buckets; i++) {
+    total_size += bucket_indices.at(i).size();
+  }
+  at::TensorOptions options;
+  options = options.dtype(at::kInt);
+  options = options.device(replicas_[0][0].device());
+  auto broadcast_tensor = at::empty({total_size}, at::kInt);
+  auto accessor = broadcast_tensor.accessor<int, 1>();
+  auto accessorIndex = 0;
+  for (size_t i = 0; i < num_buckets; i++) {
+    const auto& bucket_size = bucket_indices.at(i).size();
+    for (size_t j = 0; j < bucket_size; j++) {
+      accessor[accessorIndex++] = bucket_indices[i][j];
+    }
+  }
+  auto broadcast_tensor_device = at::empty({total_size}, options);
+  broadcast_tensor_device.copy_(broadcast_tensor, true);
+  // Copy CPU tensor to device tensor, as the process_group_ could be NCCL and
+  // it can only broadcast device tensors.
+  broadcast_coalesced(
+      process_group_, broadcast_tensor_device, broadcast_bucket_size);
+  broadcast_tensor.copy_(broadcast_tensor_device);
+  accessorIndex = 0;
+  for (size_t i = 0; i < num_buckets; i++) {
```

Comment:
Do we also need to broadcast number of buckets and size of each buckets? What if the mapping on each rank is different that results in different `num_buckets` and `bucket_size`?

Reply:
you are right!